python_modules/dagster/dagster/_core/storage/upath_io_manager.py

```diff
@@ -0,0 +1,229 @@
+from __future__ import annotations
+
+import inspect
+from abc import abstractmethod
+from typing import Any, Dict, Mapping, Union
+
+from upath import UPath
+
+from dagster import InputContext, MemoizableIOManager, MetadataValue, OutputContext
+from dagster import _check as check
+from dagster._annotations import experimental
+
+
+@experimental
+class UPathIOManagerBase(MemoizableIOManager):
+    """
+    Abstract IOManager base class compatible with local and cloud storage via `fsspec` (using `universal-pathlib`).
+
+    Features:
+     - working with any filesystem supported by `fsspec`
+     - handling loading multiple upstream asset partitions via PartitionMapping
+     (returns a dictionary with mappings from partition_keys to loaded objects)
+     - the `get_metadata` method can be customized to add additional metadata to the outputs
+     - the `allow_missing_partitions: bool` metadata value can control
+     either raising an error or skipping on missing partitions
+    """
+
+    extension: str = ""  # override in child class
+
+    @abstractmethod
+    def dump_to_path(self, context: OutputContext, obj: Any, path: UPath):
+        """
+        Child classes should override this method to write the object to the filesystem.
+
+        Args:
+            context:
+            obj:
+            path:
+
+        Returns:
+
+        """
+        pass  # pylint: disable=W0107
```

Comment:
Not a big deal either way, but FYI in the future I think you can just omit `pass` statements like this, because the docstring counts as a body of the function.